elasticsearch_dsl/search.py

```diff
@@ -15,924 +15,6 @@
-        iterate over all the documents matching the query.
-
-        Use ``params`` method to specify any additional arguments you with to
-        pass to the underlying ``scan`` helper from ``elasticsearch-py`` -
-        https://elasticsearch-py.readthedocs.io/en/master/helpers.html#elasticsearch.helpers.scan
-
-        """
-        es = get_connection(self._using)
-
-        for hit in scan(es, query=self.to_dict(), index=self._index, **self._params):
-            yield self._get_result(hit)
-
-    def delete(self):
-        """
-        delete() executes the query by delegating to delete_by_query()
-        """
-
-        es = get_connection(self._using)
-
-        return AttrDict(
-            es.delete_by_query(index=self._index, body=self.to_dict(), **self._params)
-        )
-
-
-class MultiSearch(Request):
-    """
-    Combine multiple :class:`~elasticsearch_dsl.Search` objects into a single
-    request.
-    """
-
-    def __init__(self, **kwargs):
-        super().__init__(**kwargs)
-        self._searches = []
-
-    def __getitem__(self, key):
-        return self._searches[key]
-
-    def __iter__(self):
-        return iter(self._searches)
-
-    def _clone(self):
-        ms = super()._clone()
-        ms._searches = self._searches[:]
-        return ms
-
-    def add(self, search):
-        """
-        Adds a new :class:`~elasticsearch_dsl.Search` object to the request::
-
-            ms = MultiSearch(index='my-index')
-            ms = ms.add(Search(doc_type=Category).filter('term', category='python'))
-            ms = ms.add(Search(doc_type=Blog))
-        """
-        ms = self._clone()
-        ms._searches.append(search)
-        return ms
-
-    def to_dict(self):
-        out = []
-        for s in self._searches:
-            meta = {}
-            if s._index:
-                meta["index"] = s._index
-            meta.update(s._params)
-
-            out.append(meta)
-            out.append(s.to_dict())
-
-        return out
-
-    def execute(self, ignore_cache=False, raise_on_error=True):
-        """
-        Execute the multi search request and return a list of search results.
-        """
-        if ignore_cache or not hasattr(self, "_response"):
-            es = get_connection(self._using)
-
-            responses = es.msearch(
-                index=self._index, body=self.to_dict(), **self._params
-            )
-
-            out = []
-            for s, r in zip(self._searches, responses["responses"]):
-                if r.get("error", False):
-                    if raise_on_error:
-                        raise ApiError("N/A", meta=responses.meta, body=r)
-                    r = None
-                else:
-                    r = Response(s, r)
-                out.append(r)
-
-            self._response = out
-
-        return self._response
+from elasticsearch_dsl._async.search import AsyncMultiSearch, AsyncSearch  # noqa: F401
```

Comment:
```suggestion
```

What I was trying to say in https://github.com/elastic/elasticsearch-dsl-py/pull/1714#discussion_r1531548995 is that we should *not* import the async classes here. Instead we should directly import `_async.search.AsyncSearch`  and `_async.search.AsyncMultiSearch` from the top-level `__init__.py`. My take was that all imports should be done at the top-level and we should not encourage users reaching out to subpackages as it limits our options going forward.

But maybe this is a lost cause and adding the async classes here helps users?

Reply:
I'm fine either way. I added them here because some of the tests imported this module. I'm not sure how common this is, but someone who wants to use both `Search` and `Multisearch` or both `AsyncSearch` and `AsyncMultiSearch` may want to only import `search` and access the classes from there. Is there anything wrong with that? If we do not import the async classes then the equivalent import is ugly:

```python
from elasticsearch_dsl._async import search
```

Reply:
How is `from elasticsearch_dsl.search import Search, MultiSearch` better than `from elasticsearch_dsl import Search, MultiSearch`? It's not like it would reduce the import time, right?

Reply:
Sorry, I wasn't very clear. What I'm saying is that you may want to do this:

```python
from elasticsearch_dsl import search

s = search.Search()
ms = search.MultiSearch()
```

or the async version:

```python
from elasticsearch_dsl import search

s = search.AsyncSearch()
ms = search.AsyncMultiSearch()
```

If we do not include the async classes in `search`, then the only other way to do this is with the `_async` module, which is not as nice:

```python
from elasticsearch_dsl._async import search

s = search.AsyncSearch()
ms = search.AsyncMultiSearch()
```

Reply:
I certainly don't want to encourage importing `_async`. I have seen zero evidence of users [doing what you suggest though](https://github.com/search?q=%22from+elasticsearch_dsl+import+search%22&type=code&ref=advsearch&p=1). While there's [plenty of evidence of importing through `search` for no reasons](https://github.com/search?q=%22from+elasticsearch_dsl.search+import+Search%22&type=code&ref=advsearch).

I'll let you make the decision here.

Reply:
I don't know if users do it or not, but we do it all over the place in tests:

https://github.com/elastic/elasticsearch-dsl-py/blob/main/tests/test_search.py#L22
https://github.com/elastic/elasticsearch-dsl-py/blob/main/tests/test_document.py#L32
https://github.com/elastic/elasticsearch-dsl-py/blob/main/tests/test_mapping.py#L20
https://github.com/elastic/elasticsearch-dsl-py/blob/main/tests/test_integration/test_mapping.py#L20

And I have to agree that if you are going to use a bunch of things from a module, it sometimes is more convenient to import the module instead of all the individual things.